packages/dashboard-core-plugins/src/panels/MarkdownNotebookPanel.jsx

```diff
@@ -0,0 +1,392 @@
+    this.editor = editor;
+  }
+
+  handleEditorBlur(event) {
+    log.debug(`markdown content changed, saving...`);
+    const { isEditing } = this.state;
+
+    // if not in edit mode, or in edit mode but blur went to an internal monaco field (like search)
+    if (
+      !isEditing ||
+      this.markdownEditor?.container?.contains(event.relatedTarget)
+    ) {
+      return;
+    }
+
+    const content = this.editor.getValue();
+
+    this.setState({
+      content,
+      isEditing: false,
+    });
+  }
+
+  handleEditorResize() {
+    const { isEditing } = this.state;
+    if (isEditing && this.editor) {
+      this.editor.layout();
+    }
+  }
+
+  /**
+   * @param {MouseEvent} event The click event from clicking on the link
+   */
+  handleLinkClick(event) {
+    const { notebooksUrl, session, sessionLanguage } = this.props;
+    const { href } = event.target;
+    if (!href || !href.startsWith(notebooksUrl)) {
+      return;
+    }
+
+    event.stopPropagation();
+    event.preventDefault();
+
+    const notebookPath = `/${href.substring(notebooksUrl.length)}`.replace(
+      /%20/g,
+      ' '
+    );
+    log.debug('Notebook link clicked, opening', notebookPath);
+
+    const { glEventHub } = this.props;
+    const notebookSettings = {
+      value: null,
+      language: sessionLanguage,
+    };
+    const fileMetadata = { id: notebookPath, itemName: notebookPath };
+    glEventHub.emit(
+      NotebookEvent.SELECT_NOTEBOOK,
+      session,
+      sessionLanguage,
+      notebookSettings,
+      fileMetadata
+    );
+  }
+
+  handleTab(tab) {
+    log.debug('MarkdownNotebookPanel tab event', tab);
+    this.initTab(tab);
+  }
+
+  handleRunCommand(command) {
+    if (!command) {
+      log.debug('Ignoring empty command.');
+      return;
+    }
+
+    this.setState(
+      ({ isPreview, panelState = {} }) => {
+        if (isPreview) {
+          const { metadata, panelState: propsPanelState } = this.props;
+          this.registerFileMetadata(
+            propsPanelState?.fileMetadata ?? metadata,
+            false
+          );
+          return { isPreview: false, panelState: { ...panelState, isPreview } };
+        }
+        return null;
+      },
+      () => {
+        this.setPreviewStatus();
+      }
+    );
+
+    const { glEventHub } = this.props;
+    glEventHub.emit(ConsoleEvent.SEND_COMMAND, command, false, true);
+  }
+
+  /**
+   * Transform the link URI to load from where the notebook is if it's relative
+   * @param {String} src The link to transform
+   * @returns String the transformed link
+   */
+  transformLinkUri(src) {
+    const { notebooksUrl, panelState } = this.props;
+    let itemName = panelState?.fileMetadata?.itemName;
+    if (!itemName) {
+      return src;
+    }
+    if (itemName.charAt(0) === '/') {
+      itemName = itemName.substring(1);
+    }
+
+    const itemUri = new URL(itemName, notebooksUrl);
+    if (src.charAt(0) === '/') {
+      return `${notebooksUrl}${src}`;
```

Comment:
Default `notebooksUrl` has a trailing slash already

Reply:
Does it? I don't think it does?

Reply:
Could be browser specific, I don't know... Here's what I see in Chrome for the default `notebooksUrl` prop:

```
new URL(
    `http://localhost:4000/`,
    window.location
  ).href
> 'http://localhost:4000/'
```

Reply:
I'm dumb